composer/datasets/streaming/writer.py

```diff
@@ -82,29 +88,63 @@ def __init__(self, dirname: str, fields: List[str], shard_size_limit: int = 1 <<
         self.new_samples = []
         self.new_shard_size = 0
 
+        # compression scheme for shards
+        self.compression_scheme = None
+        self.open_f = lambda n: open(n, 'xb')
```

Comment:
Instead of defining `open_f` as a lambda, could you make this a private helper function? It's a bit unstandard to define a lambda as a self variable, rather than use a function, and makes the code less readable (typically one doesn't look at `__init__` to find where methods are defined).

For example:

E.g.

```python
def _open_file(self, filename: str):
    if self.compression_scheme.startswith("gz"):
        ...
        return gzip.open(filename, ...)
    if self.compression_scheme is None:
        return open(filename, ...)
    else:
        raise ValueError(f"Invalid compression scheme: {self.compression_scheme}")
```